CPSC-483 Section: 02

Project 3

Group Members:

Divya Barsode, Devin Heng

In [87]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import sklearn

In [88]:
'''STEP 1 --> importing the CSV file'''
df = pd.read_csv('bank-additional-full.csv', sep=";")
df.head()

age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         0  nonexistent          1.1   
1   may         mon  ...         1    999         0  nonexistent          1.1   
2   may         mon  ...         1    999         0  nonexistent          1.1   
3   may         mon  ...         1    999         0  nonexistent          1.1   
4   may         mon  ...         1    999         0  nonexistent          1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0          93.994          -36.4      4.857       5191.0  no  
1          93.994          -36.4      4.857       5191.0  no  
2          93.994          -36.4      4.857       5191.0  no  
3          93.994          -36.4      4.857       5191.0  no  
4          93.994          -36.4      4.857       5191.0  no  

[5 rows x 21 columns]

In [89]:
y = df.iloc[0:, 20] #this is the output variable dataframe
X = df.iloc[0:, 0:20] #this is the rest of the factors except the output variable
#X = df.drop(['y'], axis=1)

#y = df['y']


In [90]:
'''STEP 2 --> split the features and target values 10% test 90% training'''
from sklearn.model_selection import train_test_split

#we use train_test_split to randomly generate sets and we also split the output value already
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=(2021-10-25)) # 10% test, 90% train



In [91]:
'''STEP 3 --> dropping duration column first'''

X_train = X_train.drop(columns=['duration'])
X_test = X_test.drop(columns=['duration'])

'''
X_train = X_train.drop(columns=['duration','campaign','pdays','previous','poutcome','emp.var.rate','cons.price.idx','cons.conf.idx','euribor3m','nr.employed'])
X_test = X_test.drop(columns=['duration','campaign','pdays','previous','poutcome','emp.var.rate','cons.price.idx','cons.conf.idx','euribor3m','nr.employed'])
'''


"\nX_train = X_train.drop(columns=['duration','campaign','pdays','previous','poutcome','emp.var.rate','cons.price.idx','cons.conf.idx','euribor3m','nr.employed'])\nX_test = X_test.drop(columns=['duration','campaign','pdays','previous','poutcome','emp.var.rate','cons.price.idx','cons.conf.idx','euribor3m','nr.employed'])\n"

In [92]:
'''STEP 4 --> Only account for bank client data so drop all data after column 7'''

#drop all the columns after column 7
X_train = X_train.drop(columns=['contact','month', 'day_of_week', 'campaign','pdays','previous','poutcome','emp.var.rate','cons.price.idx','cons.conf.idx','euribor3m','nr.employed'])
X_test = X_test.drop(columns=['contact','month', 'day_of_week','campaign','pdays','previous','poutcome','emp.var.rate','cons.price.idx','cons.conf.idx','euribor3m','nr.employed'])

#make the dummy variables for the categorical
#categorical vars in a list 
listOfCats = ['job','marital','education','default','housing','loan']
X_train = pd.get_dummies(data=X_train, columns=listOfCats)
X_test = pd.get_dummies(data=X_test, columns=listOfCats)


In [93]:
X_train

age  job_admin.  job_blue-collar  job_entrepreneur  job_housemaid  \
22975   56           0                0                 0              0   
14746   26           0                0                 0              0   
12505   38           1                0                 0              0   
6801    33           1                0                 0              0   
18389   29           1                0                 0              0   
...    ...         ...              ...               ...            ...   
17987   31           0                1                 0              0   
20299   29           0                0                 0              0   
14556   49           0                1                 0              0   
28981   47           1                0                 0              0   
32764   35           1                0                 0              0   

       job_management  job_retired  job_self-employed  job_services  \
22975               0            1                  0             0   
14746               0            0                  0             0   
12505               0            0                  0             0   
6801                0            0                  0             0   
18389               0            0                  0             0   
...               ...          ...                ...           ...   
17987               0            0                  0             0   
20299               0            0                  0             0   
14556               0            0                  0             0   
28981               0            0                  0             0   
32764               0            0                  0             0   

       job_student  ...  education_unknown  default_no  default_unknown  \
22975            0  ...                  0           1                0   
14746            0  ...                  0           1                0   
12505            0  ...                  0           1                0   
6801             0  ...                  0           1                0   
18389            0  ...                  0           1                0   
...            ...  ...                ...         ...              ...   
17987            0  ...                  0           1                0   
20299            0  ...                  0           1                0   
14556            0  ...                  0           0                1   
28981            0  ...                  0           1                0   
32764            0  ...                  0           1                0   

       default_yes  housing_no  housing_unknown  housing_yes  loan_no  \
22975            0           0                0            1        1   
14746            0           0                1            0        0   
12505            0           0                0            1        1   
6801             0           1                0            0        1   
18389            0           1                0            0        1   
...            ...         ...              ...          ...      ...   
17987            0           0                0            1        1   
20299            0           1                0            0        1   
14556            0           0                0            1        1   
28981            0           0                0            1        1   
32764            0           0                0            1        1   

       loan_unknown  loan_yes  
22975             0         0  
14746             1         0  
12505             0         0  
6801              0         0  
18389             0         0  
...             ...       ...  
17987             0         0  
20299             0         0  
14556             0         0  
28981             0         0  
32764             0         0  

[37069 rows x 34 columns]

In [94]:
# from sklearn.preprocessing import MinMaxScaler

# scaler = MinMaxScaler()
# X_train = scaler.fit_transform(X_train)
# X_test = scaler.fit_transform(X_test)

In [95]:
X_train = pd.DataFrame(X_train)
X_test = pd.DataFrame(X_test)

In [96]:
X_test

age  job_admin.  job_blue-collar  job_entrepreneur  job_housemaid  \
37215   38           1                0                 0              0   
576     41           1                0                 0              0   
14186   35           0                1                 0              0   
4527    26           0                1                 0              0   
3900    43           1                0                 0              0   
...    ...         ...              ...               ...            ...   
4312    41           0                0                 0              0   
27927   25           1                0                 0              0   
11635   55           1                0                 0              0   
22652   44           0                0                 0              0   
20694   54           0                1                 0              0   

       job_management  job_retired  job_self-employed  job_services  \
37215               0            0                  0             0   
576                 0            0                  0             0   
14186               0            0                  0             0   
4527                0            0                  0             0   
3900                0            0                  0             0   
...               ...          ...                ...           ...   
4312                0            0                  0             1   
27927               0            0                  0             0   
11635               0            0                  0             0   
22652               0            0                  0             0   
20694               0            0                  0             0   

       job_student  ...  education_unknown  default_no  default_unknown  \
37215            0  ...                  0           1                0   
576              0  ...                  0           1                0   
14186            0  ...                  0           0                1   
4527             0  ...                  0           1                0   
3900             0  ...                  0           1                0   
...            ...  ...                ...         ...              ...   
4312             0  ...                  0           0                1   
27927            0  ...                  0           1                0   
11635            0  ...                  0           1                0   
22652            0  ...                  0           1                0   
20694            0  ...                  0           0                1   

       default_yes  housing_no  housing_unknown  housing_yes  loan_no  \
37215            0           0                0            1        1   
576              0           1                0            0        1   
14186            0           0                0            1        0   
4527             0           0                0            1        0   
3900             0           1                0            0        1   
...            ...         ...              ...          ...      ...   
4312             0           0                0            1        1   
27927            0           0                0            1        1   
11635            0           0                0            1        1   
22652            0           0                0            1        0   
20694            0           0                0            1        1   

       loan_unknown  loan_yes  
37215             0         0  
576               0         0  
14186             0         1  
4527              0         1  
3900              0         0  
...             ...       ...  
4312              0         0  
27927             0         0  
11635             0         0  
22652             0         1  
20694             0         0  

[4119 rows x 34 columns]

In [97]:
'''STEP 5: running the data on a categorical naive bayes classifier'''
from sklearn.naive_bayes import CategoricalNB

cnb = CategoricalNB()
cnb.fit(X_train, y_train)

CategoricalNB()

In [98]:
yTrain_score = cnb.score(X_train,y_train,sample_weight=None)

In [99]:
X_test

age  job_admin.  job_blue-collar  job_entrepreneur  job_housemaid  \
37215   38           1                0                 0              0   
576     41           1                0                 0              0   
14186   35           0                1                 0              0   
4527    26           0                1                 0              0   
3900    43           1                0                 0              0   
...    ...         ...              ...               ...            ...   
4312    41           0                0                 0              0   
27927   25           1                0                 0              0   
11635   55           1                0                 0              0   
22652   44           0                0                 0              0   
20694   54           0                1                 0              0   

       job_management  job_retired  job_self-employed  job_services  \
37215               0            0                  0             0   
576                 0            0                  0             0   
14186               0            0                  0             0   
4527                0            0                  0             0   
3900                0            0                  0             0   
...               ...          ...                ...           ...   
4312                0            0                  0             1   
27927               0            0                  0             0   
11635               0            0                  0             0   
22652               0            0                  0             0   
20694               0            0                  0             0   

       job_student  ...  education_unknown  default_no  default_unknown  \
37215            0  ...                  0           1                0   
576              0  ...                  0           1                0   
14186            0  ...                  0           0                1   
4527             0  ...                  0           1                0   
3900             0  ...                  0           1                0   
...            ...  ...                ...         ...              ...   
4312             0  ...                  0           0                1   
27927            0  ...                  0           1                0   
11635            0  ...                  0           1                0   
22652            0  ...                  0           1                0   
20694            0  ...                  0           0                1   

       default_yes  housing_no  housing_unknown  housing_yes  loan_no  \
37215            0           0                0            1        1   
576              0           1                0            0        1   
14186            0           0                0            1        0   
4527             0           0                0            1        0   
3900             0           1                0            0        1   
...            ...         ...              ...          ...      ...   
4312             0           0                0            1        1   
27927            0           0                0            1        1   
11635            0           0                0            1        1   
22652            0           0                0            1        0   
20694            0           0                0            1        1   

       loan_unknown  loan_yes  
37215             0         0  
576               0         0  
14186             0         1  
4527              0         1  
3900              0         0  
...             ...       ...  
4312              0         0  
27927             0         0  
11635             0         0  
22652             0         1  
20694             0         0  

[4119 rows x 34 columns]

In [100]:
yTrain_score

0.8812484825595511

In [101]:
yTest_score = cnb.score(X_test,y_test,sample_weight=None) ## Throws error only for test
yTest_score

0.8781257586792911

In [102]:
'''STEP 6 --> How many categories are there?'''
column = df["age"]
min_value = column.min()
max_value = column.max()
max_value
min_value
#the highest age is 98 and the lowest age is 17 therefore the amount
#of categories would be 81 categories
#this would not be a reasonable size for categories

17

In [103]:
'''Step 7 --> split age into bins per decade rather than each age'''

#creating the different age group bins
ageBucketDF =  df.copy()

def ageBucket(age):
    return age//10


ageBucketDF["age"] = ageBucketDF["age"].apply(ageBucket)
ageBucketDF.head()





age        job  marital    education  default housing loan    contact  \
0    5  housemaid  married     basic.4y       no      no   no  telephone   
1    5   services  married  high.school  unknown      no   no  telephone   
2    3   services  married  high.school       no     yes   no  telephone   
3    4     admin.  married     basic.6y       no      no   no  telephone   
4    5   services  married  high.school       no      no  yes  telephone   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         0  nonexistent          1.1   
1   may         mon  ...         1    999         0  nonexistent          1.1   
2   may         mon  ...         1    999         0  nonexistent          1.1   
3   may         mon  ...         1    999         0  nonexistent          1.1   
4   may         mon  ...         1    999         0  nonexistent          1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0          93.994          -36.4      4.857       5191.0  no  
1          93.994          -36.4      4.857       5191.0  no  
2          93.994          -36.4      4.857       5191.0  no  
3          93.994          -36.4      4.857       5191.0  no  
4          93.994          -36.4      4.857       5191.0  no  

[5 rows x 21 columns]

In [104]:
#redoing the past experiments with new dataset

yBucket = ageBucketDF.iloc[0:, 20] #this is the output variable dataframe
XBucket = ageBucketDF.iloc[0:, 0:20] #this is the rest of the factors except the output variable

#we use train_test_split to randomly generate sets and we also split the output value already
X_trainBucket, X_testBucket, y_trainBucket, y_testBucket = train_test_split(XBucket, yBucket, test_size=0.1, random_state=(2021-10-25)) # 10% test, 90% train

#drop the columns that we dont need (we are still focusing on client data only)
X_trainBucket = X_trainBucket.drop(columns=['duration','contact','month', 'day_of_week', 'campaign','pdays','previous','poutcome','emp.var.rate','cons.price.idx','cons.conf.idx','euribor3m','nr.employed'])
X_testBucket = X_testBucket.drop(columns=['duration','contact','month', 'day_of_week', 'campaign','pdays','previous','poutcome','emp.var.rate','cons.price.idx','cons.conf.idx','euribor3m','nr.employed'])

#creating our dummy variables again
X_trainBucket = pd.get_dummies(data=X_trainBucket, columns=listOfCats)
X_testBucket = pd.get_dummies(data=X_testBucket, columns=listOfCats)

'''
not sure if we need to scaler fit for this
X_trainBucket = scaler.fit_transform(X_trainBucket)
X_testBucket = scaler.fit_transform(X_testBucket)
'''
#making them into pandas dataframes
X_trainBucket = pd.DataFrame(X_trainBucket)
X_testBucket = pd.DataFrame(X_testBucket)

#running the Naive Bayes on the new data
cnbBucket = CategoricalNB()
cnbBucket.fit(X_trainBucket, y_trainBucket)


CategoricalNB()

In [105]:
yTrain_scoreBucket = cnbBucket.score(X_trainBucket,y_trainBucket,sample_weight=None)
yTrain_scoreBucket

0.8809787153686368

In [106]:
yTest_scoreBucket = cnbBucket.score(X_testBucket,y_testBucket,sample_weight=None) ## Throws error only for test
yTest_scoreBucket

0.8783685360524399

In [107]:
'''STEP 7 Result: It seems like the performance does not really change
with the ages put into decades rather than the exact age'''

'STEP 7 Result: It seems like the performance does not really change\nwith the ages put into decades rather than the exact age'

In [108]:
'''Step 8 --> Run experiment 5(with original data) on KNN classifier'''
from sklearn.neighbors import (NeighborhoodComponentsAnalysis,KNeighborsClassifier)
from sklearn import metrics

knn = KNeighborsClassifier(n_neighbors=6)
knn.fit(X_train, y_train)



KNeighborsClassifier(n_neighbors=6)

In [109]:
yTrain_scoreKnn = knn.score(X_train,y_train)
#k = 6 --> 89%
#k = 3 --> 88
print(yTrain_scoreKnn)

0.8913647522188352


In [110]:
ytest_scoreKnn = knn.score(X_test,y_test)
#k = 6 --> 88%
#k = 3 --> 86%
ytest_scoreKnn

0.883709638261714

In [111]:
'''STEP 9 --> how many responses are 0 and 1 in test set'''
y_test.value_counts()
#there are 3646 no's and 473 yes's

no     3646
yes     473
Name: y, dtype: int64

In [155]:
'''Step 10 --> checking to see what would happen if we got all 0 as predicted'''
from sklearn.metrics import confusion_matrix
zerosBin = np.zeros_like(y_test)  #we need binary for the AUC score
zerosClass = np.zeros_like(y_test) #we need in terms of 'no' for conf. matrix

zerosClass[zerosClass==0] = 'no'

y_testScore = knn.score(X_test,zerosClass)

confusionMatrix = confusion_matrix(y_test, zerosClass) ## What to do of Tar??? what to send in the second param

print("Confusion Matrix: \n",confusionMatrix)


Confusion Matrix: 
 [[3646    0]
 [ 473    0]]


In [158]:
from sklearn.metrics import roc_auc_score
zerosClass1 = np.zeros_like(y_test)
zerosClass1 = np.zeros_like(y_test)
aucTest = roc_auc_score(y_test, zerosClass1)
print("AUC:",aucTest)

AUC: 0.5


In [128]:
''' Step 11 --> confusion matrix and AUC for steps 7 and 8'''
#################
y_predTrainAgeBucket = cnb.predict(X_testBucket)

confusionMatrix = confusion_matrix(y_testBucket, y_predTrainAgeBucket) #pass in true value and predicted values
print("Confusion Matrix: \n",confusionMatrix)

aucTrainBucket = roc_auc_score(y_trainBucket, cnb.predict_proba(X_trainBucket)[:, 1])
aucTrainBucket

Confusion Matrix: 
 [[1845 1801]
 [ 150  323]]


0.6462782217590364

### The confusion matrix produces this
 [[1845 1801 ]
 
 [ 150  323]]
 
 TP = 1845
 FP = 1801
 FN = 150
 TN = 323
 
 From this information we can say the classifier isn't performing that well and is only getting roughly 50 percent of the predictions correct on the testing set.
 

In [ ]:
'''STEP 12 --> fixing imbalanced data'''
#https://machinelearningmastery.com/random-oversampling-and-undersampling-for-imbalanced-classification/

#start from the original dataframe 
balancedDf = df.copy()

#get the amount of 0's and amount of 1's
zeroBalanced = balancedDf['y'].value_counts()['no'] #amount of 0
oneBalanced = balancedDf['y'].value_counts()['yes'] #amount of 1

#setting the weights for the sample
weights = {'no': oneBalanced,
           'yes': zeroBalanced}

#create a new column that assigns the weights to the corresponding response
balancedDf['weights'] = balancedDf['y'].apply(lambda x: weights[x])

balanced = balancedDf.sample(frac=1.0, replace=True, random_state=(2021-10-25), weights = 'weights')

print(len(balanced))
print(balanced['y'].value_counts()['no'])
print(balanced['y'].value_counts()['yes'])
#the amounts become our weights
#weight of no = zeroBalanced...weight of yes = oneBalanced
#we need to inverse these and give 1 the weight of 0 and vice versa
#create new column -> where y is 'no' the weight is oneBalanced value
    #else the value of 'yes' becomes zeroBalanced



In [120]:
'''STEP 13 --> Using the balanced set...now redo the training steps and see what we get'''
balanced_y = balanced.iloc[0:, 20] #this is the output variable dataframe
balanced_X = balanced.iloc[0:, 0:20] #this is the rest of the factors except the output variable

#there is only one 'default yes' so we get rid of it to see how it fits
#balanced_X.loc[(balanced_X.default == 'yes'), 'default'] = 'unknown'

#from the balanced X and Y --> get balanced training/testing sets
X_bal_train, X_bal_test, y_bal_train, y_bal_test = train_test_split(balanced_X, balanced_y, test_size=0.1, random_state=(2021-10-25)) # 10% test, 90% train


#drop the duration column from the sets
X_bal_train = X_bal_train.drop(columns=['duration'])
X_bal_test = X_bal_test.drop(columns=['duration'])

#only focus on bank client data and make dummy variables
#drop all the columns after column 7
X_bal_train = X_bal_train.drop(columns=['contact','month', 'day_of_week', 'campaign','pdays','previous','poutcome','emp.var.rate','cons.price.idx','cons.conf.idx','euribor3m','nr.employed'])
X_bal_test = X_bal_test.drop(columns=['contact','month', 'day_of_week','campaign','pdays','previous','poutcome','emp.var.rate','cons.price.idx','cons.conf.idx','euribor3m','nr.employed'])

#make the dummy variables for the categorical
#categorical vars in a list 
listOfCats = ['job','marital','education','default','housing','loan']

X_bal_train = pd.get_dummies(data=X_bal_train, columns=listOfCats)
X_bal_test = pd.get_dummies(data=X_bal_test, columns=listOfCats)

#turn the bal with dummies into a dataframe
X_bal_train = pd.DataFrame(X_bal_train)
X_bal_test = pd.DataFrame(X_bal_test)

X_bal_test['default_yes'] = 0
X_bal_train.to_csv('train.csv')
X_bal_test.to_csv('test.csv')



In [121]:
X_bal_test

age  job_admin.  job_blue-collar  job_entrepreneur  job_housemaid  \
40996   81           0                0                 0              0   
37068   32           1                0                 0              0   
12983   30           0                1                 0              0   
11322   46           0                0                 1              0   
20702   29           0                0                 0              0   
...    ...         ...              ...               ...            ...   
10924   28           1                0                 0              0   
16278   27           1                0                 0              0   
12507   55           0                1                 0              0   
30993   48           0                1                 0              0   
16884   34           0                1                 0              0   

       job_management  job_retired  job_self-employed  job_services  \
40996               0            1                  0             0   
37068               0            0                  0             0   
12983               0            0                  0             0   
11322               0            0                  0             0   
20702               0            0                  0             0   
...               ...          ...                ...           ...   
10924               0            0                  0             0   
16278               0            0                  0             0   
12507               0            0                  0             0   
30993               0            0                  0             0   
16884               0            0                  0             0   

       job_student  ...  education_unknown  default_no  default_unknown  \
40996            0  ...                  0           1                0   
37068            0  ...                  0           1                0   
12983            0  ...                  0           1                0   
11322            0  ...                  0           0                1   
20702            0  ...                  0           1                0   
...            ...  ...                ...         ...              ...   
10924            0  ...                  0           1                0   
16278            0  ...                  0           1                0   
12507            0  ...                  0           1                0   
30993            0  ...                  0           1                0   
16884            0  ...                  0           1                0   

       housing_no  housing_unknown  housing_yes  loan_no  loan_unknown  \
40996           0                0            1        1             0   
37068           0                0            1        1             0   
12983           1                0            0        1             0   
11322           0                0            1        1             0   
20702           1                0            0        1             0   
...           ...              ...          ...      ...           ...   
10924           0                0            1        1             0   
16278           1                0            0        1             0   
12507           1                0            0        1             0   
30993           0                0            1        1             0   
16884           0                0            1        1             0   

       loan_yes  default_yes  
40996         0            0  
37068         0            0  
12983         0            0  
11322         0            0  
20702         0            0  
...         ...          ...  
10924         0            0  
16278         0            0  
12507         0            0  
30993         0            0  
16884         0            0  

[4119 rows x 34 columns]

In [129]:
'''STEP 13 --> Using the balanced set...now redo the training steps and see what we get'''
#categorical naive bayes training score
cnb_balanced = CategoricalNB()
cnb_balanced.fit(X_bal_train, y_bal_train)
yTrain_bal_score = cnb_balanced.score(X_bal_train,y_bal_train,sample_weight=None)
yTrain_bal_score


0.6129110577571556

In [130]:
'''STEP 13 --> Using the balanced set...now redo the training steps and see what we get'''
#test set on the model
yTest_bal_score = cnb_balanced.score(X_bal_test,y_bal_test,sample_weight=None) ## Throws error only for test
yTest_bal_score

0.5962612284535082

In [131]:
'''STEP 13 --> Using the balanced set...now redo the training steps and see what we get'''
#confusion matrix and AUC of the NB classifier
y_bal_predTrain = cnb_balanced.predict(X_bal_train)

confusionMatrix = confusion_matrix(y_bal_train, y_bal_predTrain) #pass in true value and predicted values
print("Confusion Matrix: \n",confusionMatrix)

auc = roc_auc_score(y_bal_train, cnb_balanced.predict_proba(X_bal_train)[:, 1])
auc

Confusion Matrix: 
 [[10503  7956]
 [ 6393 12217]]


0.6650963523470506

In [132]:
'''STEP 13 --> Using the balanced set...now redo the training steps and see what we get'''
#Now run the balanced dataset on the KNN classifier
knn_balanced = KNeighborsClassifier(n_neighbors=6)
knn_balanced.fit(X_bal_train, y_bal_train)

yTrain_scoreKnn = knn_balanced.score(X_bal_train,y_bal_train)
yTrain_scoreKnn

0.7640346381073134

In [133]:
'''STEP 13 --> Using the balanced set...now redo the training steps and see what we get'''
#test data on the trained KNN classifier
ytest_scoreKnn = knn_balanced.score(X_bal_test,y_bal_test)

ytest_scoreKnn

0.6392328235008498

In [148]:
'''STEP 13 --> Using the balanced set...now redo the training steps and see what we get'''
#confusion matrix and AUC of the NB classifier
y_bal_predTrain = knn_balanced.predict(X_bal_test)

confusionMatrix = confusion_matrix(y_bal_test, y_bal_predTrain) #pass in true value and predicted values
print("Confusion Matrix: \n",confusionMatrix)

auc = roc_auc_score(y_bal_test, knn_balanced.predict_proba(X_bal_test)[:, 1])
auc

Confusion Matrix: 
 [[1449  624]
 [ 862 1184]]


0.6945286155990605

In [143]:
'''STEP 14 --> Using Economical and social attributes with Gaussian Naive Bayes '''
#creating a dataframe with only socioeconomic features
socioEconomicDf =  df.copy()
socioEconomic = socioEconomicDf.filter(['emp.var.rate', 'cons.price.idx','cons.conf.idx','euribor3m','nr.employed','y']) 
y_socioEconomic = socioEconomic.iloc[0:, 5] #this is the output variable dataframe
X_socioEconomic = socioEconomic.iloc[0:, 0:5]

#splitting into training and testing
X_trainSocioEconomic, X_testSocioEconomic, y_trainSocioEconomic, y_testSocioEconomic = train_test_split(X_socioEconomic, y_socioEconomic, test_size=0.1, random_state=(2021-10-25)) # 10% test, 90% train
X_trainSocioEconomic = pd.DataFrame(X_trainSocioEconomic)
X_testSocioEconomic = pd.DataFrame(X_testSocioEconomic)



array(['no', 'no', 'no', ..., 'no', 'no', 'no'], dtype=object)

In [147]:
'''STEP 14 --> Using Economical and social attributes with Gaussian Naive Bayes '''
from sklearn.naive_bayes import GaussianNB
#running the Gaussian Naive Bayes on this new data
gnbEcon = GaussianNB()
gnbEcon.fit(X_trainSocioEconomic, y_trainSocioEconomic)
yTrainGB_score = gnbEcon.score(X_trainSocioEconomic,y_trainSocioEconomic,sample_weight=None) ## Throws error only for test
print("training score:",yTrainGB_score)
yTestGNB_score = gnbEcon.score(X_testSocioEconomic,y_testSocioEconomic,sample_weight=None) ## Throws error only for test
print("test score:",yTestGNB_score)





training score: 0.7199816558310178
test score: 0.7193493566399611


In [149]:
'''STEP 14 --> Using Economical and social attributes with Gaussian Naive Bayes '''
predicted = gnbEcon.predict(X_testSocioEconomic)
confusionMatrix1 = confusion_matrix(y_testSocioEconomic, predicted) 

print("Confusion Matrix: \n",confusionMatrix1)

AUCSOCIO = roc_auc_score(y_testSocioEconomic, gnbEcon.predict_proba(X_testSocioEconomic)[:, 1])
print("AUC for this:",AUCSOCIO)



Confusion Matrix: 
 [[2634 1012]
 [ 144  329]]
AUC for this: 0.7348436526924581


In [ ]:
'''STEP 15 --> balance the dataset and do it again'''

In [154]:
#we have balanced_y and balanced_X already

balanced_X = balanced_X.filter(['emp.var.rate', 'cons.price.idx','cons.conf.idx','euribor3m','nr.employed','y']) 

#splitting into training and testing
X_bal_trainSocioEconomic, X_bal_testSocioEconomic, y_bal_trainSocioEconomic, y_bal_testSocioEconomic = train_test_split(balanced_X, balanced_y, test_size=0.1, random_state=(2021-10-25)) # 10% test, 90% train
X_bal_trainSocioEconomic = pd.DataFrame(X_bal_trainSocioEconomic)
X_bal_testSocioEconomic = pd.DataFrame(X_bal_testSocioEconomic)

#running the Gaussian Naive Bayes on this new data
gnbBalEcon = GaussianNB()
gnbBalEcon.fit(X_bal_trainSocioEconomic, y_bal_trainSocioEconomic)
yTrainGB_score = gnbEcon.score(X_bal_trainSocioEconomic,y_bal_trainSocioEconomic,sample_weight=None) ## Throws error only for test
print("training score:",yTrainGB_score)
yTestGNB_score = gnbEcon.score(X_bal_testSocioEconomic,y_bal_testSocioEconomic,sample_weight=None) ## Throws error only for test
print("test score:",yTestGNB_score)

#confusion matrix and AUC
predicted = gnbBalEcon.predict(X_bal_testSocioEconomic)
confusionMatrix1 = confusion_matrix(y_bal_testSocioEconomic, predicted) 

print("Confusion Matrix: \n",confusionMatrix1)

AUCSOCIO = roc_auc_score(y_bal_testSocioEconomic, gnbBalEcon.predict_proba(X_bal_testSocioEconomic)[:, 1])
print("AUC for this:",AUCSOCIO)



training score: 0.7193881680110065
test score: 0.7103665938334547
Confusion Matrix: 
 [[1466  607]
 [ 586 1460]]
AUC for this: 0.745399232981512
